File: figments/src/liber8tion/rhythm.c

/// Generates a 16-bit "sawtooth" wave at a given BPM, with BPM
/// specified in Q8.8 fixed-point format.
/// @param beats_per_minute_88 the frequency of the wave, in Q8.8 format
/// @param timebase the time offset of the wave from the millis() timer
/// @warning The BPM parameter **MUST** be provided in Q8.8 format! E.g.
/// for 120 BPM it would be 120*256 = 30720. If you just want to specify
/// "120", use beat16() or beat8().
LIB8STATIC uint16_t beat88( accum88 beats_per_minute_88, uint32_t timebase = 0)
{
    // BPM is 'beats per minute', or 'beats per 60000ms'.
    // To avoid using the (slower) division operator, we
    // want to convert 'beats per 60000ms' to 'beats per 65536ms',
    // and then use a simple, fast bit-shift to divide by 65536.
    //
    // The ratio 65536:60000 is 279.620266667:256; we'll call it 280:256.
    // The conversion is accurate to about 0.05%, more or less,
    // e.g. if you ask for "120 BPM", you'll get about "119.93".
    return (((GET_MILLIS()) - timebase) * beats_per_minute_88 * 280) >> 16;
}

/// Generates a 16-bit "sawtooth" wave at a given BPM
/// @param beats_per_minute the frequency of the wave, in decimal
/// @param timebase the time offset of the wave from the millis() timer
LIB8STATIC uint16_t beat16( accum88 beats_per_minute, uint32_t timebase = 0)
{
    // Convert simple 8-bit BPM's to full Q8.8 accum88's if needed
    if( beats_per_minute < 256) beats_per_minute <<= 8;
    return beat88(beats_per_minute, timebase);
}

/// Generates an 8-bit "sawtooth" wave at a given BPM
/// @param beats_per_minute the frequency of the wave, in decimal
/// @param timebase the time offset of the wave from the millis() timer
LIB8STATIC uint8_t beat8( accum88 beats_per_minute, uint32_t timebase = 0)
{
    return beat16( beats_per_minute, timebase) >> 8;
}


/// Generates an 8-bit sine wave at a given BPM that oscillates within
/// a given range.
/// @param beats_per_minute the frequency of the wave, in decimal
/// @param lowest the lowest output value of the sine wave
/// @param highest the highest output value of the sine wave
/// @param timebase the time offset of the wave from the millis() timer
/// @param phase_offset phase offset of the wave from the current position
LIB8STATIC uint8_t beatsin8( accum88 beats_per_minute, uint8_t lowest = 0, uint8_t highest = 255,
    uint32_t timebase = 0, uint8_t phase_offset = 0)
{
uint8_t beat = beat8( beats_per_minute, timebase);
uint8_t beatsin = sin8( beat + phase_offset);
uint8_t rangewidth = highest - lowest;
uint8_t scaledbeat = scale8( beatsin, rangewidth);
uint8_t result = lowest + scaledbeat;
return result;
}

File: figments/src/liber8tion/rhythm.rs
use num::traits::WrappingAdd;

use crate::liber8tion::{interpolate::Fract8, trig::Trig8};


fn beat88(now: u32, bpm: u16, timebase: u32) -> u16 {
    (((now - timebase).wrapping_mul(bpm as u32).wrapping_mul(280)).wrapping_shr(16)) as u16
}

fn beat16(now: u32, bpm: u16, timebase: u32) -> u16 {
    let adj_bpm = if bpm < 256 {
        bpm.wrapping_shl(8)
    } else {
        bpm
    };
    beat88(now, adj_bpm, timebase)
}

fn beat8(now: u32, bpm: u16, timebase: u32) -> Fract8 {
    Fract8::from_raw(beat16(now, bpm, timebase).wrapping_shr(8) as u8)
}

pub fn beatsin8(now: u32, bpm: u16, lowest: Fract8, highest: Fract8, timebase: u32, phase: Fract8) -> Fract8 {
    let beat = beat8(now, bpm, timebase);
    let beatsin = beat.wrapping_add(&phase).sin8();
    let width = highest - lowest;
    let scaledbeat = beatsin * width;
    
    lowest + scaledbeat
} 